File: examples/quickstart/typescript/my_client.ts
import { HttpTransport, Client } from './pulserpc/index.js';

const port = parseInt(process.env.SERVER_PORT || '8080', 10);
const transport = new HttpTransport(`http://127.0.0.1:${port}`);
const client = await Client.create(transport);

async function main() {
  // Client.create() fetches IDL from server and creates interface proxies
  const products = await client.CatalogService.listProducts();
  console.log('=== Products ===');
  for (const p of products) {
    console.log(`${p.name} - $${p.price}`);
  }

  const result = await client.CartService.addToCart({
    cartId: null,
    productId: products[0].productId,
    quantity: 2
  });
  console.log(`\nCart: ${result.cartId}`);

  const response = await client.OrderService.createOrder({
    cartId: result.cartId,
    shippingAddress: {
      street: '123 Main St',
      city: 'San Francisco',
      state: 'CA',
      zipCode: '94105',
      country: 'USA'
    },
    paymentMethod: 'credit_card'
  });
  console.log(`✓ Order created: ${response.orderId}`);

  // Test error case: empty cart
  console.log('\n=== Testing Error Case ===');
  await client.CartService.clearCart(result.cartId);
  try {
    await client.OrderService.createOrder({
      cartId: result.cartId,
      shippingAddress: {
        street: '123 Main St',
        city: 'San Francisco',
        state: 'CA',
        zipCode: '94105',
        country: 'USA'
      },
      paymentMethod: 'credit_card'
    });
    console.log('✗ Should have failed!');
  } catch (e: any) {
    console.log(`✓ Got expected error: ${e.code} - ${e.message}`);
  }
}

main().catch(console.error);

File: examples/quickstart/python/my_client.py
#!/usr/bin/env python3
import os
from pulserpc import HttpTransport, Client, RPCError

# Connect to server
port = os.environ.get("SERVER_PORT", "8080")
transport = HttpTransport(f"http://localhost:{port}")
client = Client(transport)

# List products
print("=== Products ===")
products = client.CatalogService.listProducts()
for p in products:
    print(f"{p['name']} - ${p['price']:.2f}")

# Create cart and add items
print("\n=== Creating Cart ===")
cart_response = client.CartService.addToCart({
    'productId': products[0]['productId'],
    'quantity': 2
})
my_cart = cart_response
print(f"Cart: {my_cart['cartId']}, Subtotal: ${my_cart['subtotal']:.2f}")

# Add another item
cart_response = client.CartService.addToCart({
    'cartId': my_cart['cartId'],
    'productId': products[1]['productId'],
    'quantity': 1
})
my_cart = cart_response
print(f"Updated Subtotal: ${my_cart['subtotal']:.2f}")

# Create order
print("\n=== Creating Order ===")
try:
    response_data = client.OrderService.createOrder({
        'cartId': my_cart['cartId'],
        'shippingAddress': {
            'street': '123 Main St',
            'city': 'San Francisco',
            'state': 'CA',
            'zipCode': '94105',
            'country': 'USA'
        },
        'paymentMethod': 'credit_card'
    })
    response = response_data
    print(f"✓ Order created: {response['orderId']}")
except RPCError as e:
    print(f"✗ Error {e.code}: {e.message}")

# Test error case: empty cart
print("\n=== Testing Error Case ===")
client.CartService.clearCart(my_cart['cartId'])
try:
    client.OrderService.createOrder({
        'cartId': my_cart['cartId'],
        'shippingAddress': {
            'street': '123 Main St',
            'city': 'San Francisco',
            'state': 'CA',
            'zipCode': '94105',
            'country': 'USA'
        },
        'paymentMethod': 'credit_card'
    })
    print("✗ Should have failed!")
except RPCError as e:
    print(f"✓ Got expected error: {e.code} - {e.message}")
